### PR TITLE
Adds code to train word2vec on PubMed

### PR DESCRIPTION
**Pull request by rodrigo-eai**
Implements [ch160].

### CODE REVIEW

File: LICENSE [hunk 1]
```diff
@@ -1,11 +1,11 @@
...
 
-Copyright (c) 2019 Authors of Tshitoyan et al. Nature (2019)
+Copyright(c) 2019 Authors of Tshitoyan et al. Nature(2019)
 
 Permission is hereby granted, free of charge, to any person obtaining a copy
-of this software and associated documentation files (the "Software"), to deal
+of this software and associated documentation files(the "Software"), to deal
 in the Software without restriction, including without limitation the rights
-to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
```
Comment (ghost) [inline LICENSE]: I would refrain from changing their license file, or generally to making any unnecessary changes to their code that we don't intend to upstream.

File: README.md [hunk 2]
```diff
@@ -1,109 +1,56 @@
-### Supplementary Materials for "Unsupervised word embeddings capture latent knowledge from materials science literature", *Nature*  571, 95–98 (2019).
-#### Tshitoyan, V., Dagdelen, J., Weston, L., Dunn, A., Rong, Z., Kononova, O., Persson, K. A., Ceder, G. and Jain, A. 
-doi: [10.1038/s41586-019-1335-8](https://www.nature.com/articles/s41586-019-1335-8)
```
Comment (ghost) [inline README.md]: I wouldn't remove these lines.

**Review by ghost — CHANGES_REQUESTED** [review]

File: LICENSE (hunk 1 above)
Comment (rodrigo-eai) [reply LICENSE]: Sorry, autopep8 was applied to this file by mistake.

File: mat2vec/processing/pubmed_to_lines.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+
+    Args:
+        args: A tuple containing the input and output file paths.
+        text_processor: an instance of process.MaterialsTextProcessor.
+
+    Returns:
+        doc_count: number of abstracts written to the file.
+    """
+    input_file, output_file = args
+    fout = open(output_file, 'w')
```
Comment (emily-eai) [inline mat2vec/processing/pubmed_to_lines.py:37]: Maybe using `with open(output_file, 'w') as fout:` might be cleaner? no need `fout.close()` in that case.

File: mat2vec/processing/pubmed_to_lines.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+                output_folder, file_name.replace('.xml.gz', '.txt')))
+
+    logging.info('{} files found'.format(len(input_files)))
+    text_processor = process.MaterialsTextProcessor()
+    abstracts_to_lines_fn_partial = partial(
+        abstracts_to_lines_fn,
+        text_processor=text_processor)
+
+    # Warning: more than 8 processes causes OOM on a 64GB machine.
+    pool = multiprocessing.Pool(n_processes)
```
Comment (emily-eai) [inline mat2vec/processing/pubmed_to_lines.py:85]: Can also be in the format `with multiprocessing.Pool...`

File: mat2vec/training/evaluate.py [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+
+logging.basicConfig(format="%(asctime)s : %(levelname)s : %(message)s",
+                    level=logging.INFO)
+
+
+def find_entity(entities: str, vocab: Set[str]) -> str:
+    """Returns the first match of a list of entities in a vocabulary.
+
+    Args:
+        entities: A list of entities.
```
Comment (emily-eai) [inline mat2vec/training/evaluate.py]: Says a `list` here but the input is a `str`. I understand that the list is generated inside from the string but it seems confusing.

File: mat2vec/training/evaluate.py [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+        vocab: A set of ngrams.
+
+    Returns:
+        entity: The first match of `entities` in `vocab`. Returns `None`
+            if there was no match.
+    """
+    for entity in entities.split('|'):
+        entity = entity.strip()
+        if entity in vocab:
+            return entity
```
Comment (emily-eai) [inline mat2vec/training/evaluate.py:37]: Are the entities supposed to be related or very similar? Is that why as long as the first entity is found the rest don't need to be compared?

File: mat2vec/training/evaluate.py [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+
+    Args:
+        file_name: Path to the json-lines files from Pubtator.
+        types: List of entity types to be loaded.
+
+    Returns:
+        entities: A dictionary where key is an entity type and value is a
+            set of entities of that type.
+    """
+    entities = collections.defaultdict(set)
```
Comment (emily-eai) [inline mat2vec/training/evaluate.py]: Here, `entities` is a dictionary. Maybe name them differently to differentiate the variables.

File: mat2vec/training/evaluate.py [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+            json_obj = json.loads(line)
+            for entity in json_obj['entity_list']:
+                if entity['type'].lower() in types:
+                    for entity_i in entity['name'].lower().split('|'):
+                        entities[entity['type'].lower()].add(entity_i.strip())
+    return entities
+
+
+def load_entities(folder: str, types: List[str]) -> List[Set[str]]:
+    """Given a folder with json-lines files from Pubtator, loads the entities.
```
Comment (emily-eai) [inline mat2vec/training/evaluate.py:65]: Is it okay to have the same docstring with different input arguments? Maybe describe the `fn` one as a helper function?

File: mat2vec/processing/pubmed_to_lines.py [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+        for abstract_text in root.iterfind('.//AbstractText'):
+            doc_text = abstract_text.text
+            if doc_text is None:
+                continue
+            doc_tokens, _ = text_processor.process(doc_text)
+            doc_text = ' '.join(doc_tokens)
+
+            fout.write('{}\n'.format(doc_text))
+            doc_count += 1
+            # if doc_count >= 1000:
```
Comment (emily-eai) [inline mat2vec/processing/pubmed_to_lines.py]: Should probably be removed

File: mat2vec/processing/pubmed_to_lines.py [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+        output_folder: Path to the folder to write the raw text files.
+        n_processes: Number of processes to read and write the files in
+            parallel.
+    """
+
+    start_time = time.time()
+    input_files = []
+    output_files = []
+    for file_name in os.listdir(input_folder):
+        if file_name.endswith('.gz'):
```
Comment (emily-eai) [inline mat2vec/processing/pubmed_to_lines.py:73]: Are we assuming that the files should be in .gz? I don't think this is previously mentioned.

File: mat2vec/processing/pubmed_to_lines.py [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+        input_folder: Path to the folder containing the PubMed XML files.
+        output_folder: Path to the folder to write the raw text files.
+        n_processes: Number of processes to read and write the files in
+            parallel.
+    """
+
+    start_time = time.time()
+    input_files = []
+    output_files = []
+    for file_name in os.listdir(input_folder):
```
Comment (emily-eai) [inline mat2vec/processing/pubmed_to_lines.py:72]: Since the files share the same basename, maybe extract the basename separately and then add the directories and extensions for clarity.
```suggestion
    files = [os.path.splitext(x)[0] for x in os.listdir(input_folder) if x.endswith('.xml.gz')]
    input_files = [os.path.join(input_folder, x + '.xml.gz') for x in files]
    output_files = [os.path.join(output_folder, x + '.txt') for x in files]
```

File: mat2vec/processing/pubmed_to_lines.py [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+        args: A tuple containing the input and output file paths.
+        text_processor: an instance of process.MaterialsTextProcessor.
+
+    Returns:
+        doc_count: number of abstracts written to the file.
+    """
+    input_file, output_file = args
+    fout = open(output_file, 'w')
+    doc_count = 0
+    with gzip.open(input_file, 'rt') as f:
```
Comment (emily-eai) [inline mat2vec/processing/pubmed_to_lines.py]: I think you assume that the file would be gzipped everywhere, but the documentations only refer to XML files, so it's inconsistent.

File: mat2vec/processing/pubmed_to_lines.py [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+        text = f.read()
+        tree = ET.ElementTree(ET.fromstring(text))
+        root = tree.getroot()
+
+        for abstract_text in root.iterfind('.//AbstractText'):
+            doc_text = abstract_text.text
+            if doc_text is None:
+                continue
+            doc_tokens, _ = text_processor.process(doc_text)
+            doc_text = ' '.join(doc_tokens)
```
Comment (emily-eai) [inline mat2vec/processing/pubmed_to_lines.py]: Just to clarify, the PubMed XML file you are referring to contains multiple PubMed articles, which you refer to as `docs`. And you are assuming that each `doc` contains a single `abstract` from the tag `AbstractText`. This isn't always true (see [https://www.ncbi.nlm.nih.gov/pubmed/24232444?report=xml&format=text](url). For me, I had to do this for each article (`pmid_xml` is XML file for one article only)

```
    article_info['abstract'] = ' '.join([ element_text(x) for x in list(pmid_xml.iter('AbstractText')) ])
```
where I also processed the text first to remove HTML tags that appear sometimes in different parts of the XML.

```
def element_text(element: ET.Element) -> str:
    """Convert element to text string in case it consists HTML tags that
    are being parsed as XML tags."""
    return ET.tostring(element, encoding='unicode', method='text').strip()
```

File: mat2vec/training/evaluate.py [hunk 14]
```diff
@@ -0,0 +1,201 @@
...
+
+    Args:
+        folder: Path a folder containg json-lines files from Pubtator.
```
Comment (emily-eai) [inline mat2vec/training/evaluate.py:68]: Some typos here.

File: mat2vec/training/evaluate.py [hunk 15]
```diff
@@ -0,0 +1,201 @@
...
+    Args:
+        file_name: Path to the json-lines files from Pubtator.
+        types: List of entity types to be loaded.
+
+    Returns:
+        entities: A dictionary where key is an entity type and value is a
+            set of entities of that type.
+    """
+    entities = collections.defaultdict(set)
+    with open(file_name) as f:
```
Comment (emily-eai) [inline mat2vec/training/evaluate.py:53]: Might be helpful to have a brief description of how the input json file is structured.

File: mat2vec/training/evaluate.py [hunk 16]
```diff
@@ -0,0 +1,201 @@
...
+                        help='Path to the pretrained word2vec model.')
+    parser.add_argument('--ref_pairs', required=True,
+                        help='Reference file containing the entity pairs '
+                             'separated by [SEP].')
+    parser.add_argument(
+        '--bioconcepts_path',
+        required=True,
+        help='BioConcepts file containing the entity and types.')
+    parser.add_argument('--output',
+                        help='Output file to write the top retrived words '
```
Comment (emily-eai) [inline mat2vec/training/evaluate.py:190]: Some typos here and in a couple of lines below.  **retrived** and **Do do write**

File: mat2vec/training/helpers/utils.py [hunk 17]
```diff
@@ -14,16 +14,36 @@
...
+    ")",
+    "<",
+    ">",
+    "\"",
+    "“",
+    "”",
+    "≥",
+    "≤",
+    "<nUm>"]
+EXCLUDE_TERMS = ["=", ".", ",", "(", ")", "<", ">", "\"", "“", "”", "≥", "≤",
```
Comment (emily-eai) [inline mat2vec/training/helpers/utils.py:34]: Are these terms from a vocabulary list somewhere? Could also add terms like **study** and **studies**.

File: mat2vec/processing/pubmed_to_lines.py [hunk 18]
```diff
@@ -0,0 +1,133 @@
+"""Converts PubMed abstracts to a file with one abstract per line."""
+import argparse
+import datetime
+import gzip
+import logging
```
Comment (emily-eai) [inline mat2vec/processing/pubmed_to_lines.py:5]: Have we decided if we should use loguru.logger or logging? I have switched all of mine to loguru but we should probably all be using the same thing.

File: mat2vec/processing/pubmed_to_lines.py [hunk 19]
```diff
@@ -0,0 +1,133 @@
...
+        datetime.timedelta(seconds=time.time() - start_time)))
+
+
+if __name__ == '__main__':
+
+    parser = argparse.ArgumentParser(
+        description='Converts PubMed abstracts to a file with one abstract '
+                    'per line.')
+    parser.add_argument('--input_folder', required=True,
+                        help='Folder containing the XML.gz files.')
```
Comment (emily-eai) [inline mat2vec/processing/pubmed_to_lines.py:112]: I see that you mention it here, but it's the only instance and should probably be documented in the function itself since it assumes that.

File: mat2vec/training/evaluate.py [hunk 20]
```diff
@@ -0,0 +1,201 @@
...
+    pool = multiprocessing.Pool()
+    load_entities_fn_partial = partial(load_entities_fn, types=types)
+
+    results = pool.map(load_entities_fn_partial, file_names)
+    pool.close()
+    pool.join()
+
+    # Merge sets.
+    entities = {type_: set() for type_ in types}
+    entities = collections.defaultdict(set)
```
Comment (emily-eai) [inline mat2vec/training/evaluate.py:87]: Duplicate initializations of entities?

File: mat2vec/training/evaluate.py [hunk 21]
```diff
@@ -0,0 +1,201 @@
...
+    entities2_found = 0
+    unique_entities1 = set()
+    unique_entities2 = set()
+    pairs_found = 0
+    pairs_total = 0
+    recall = 0
+    MRR = 0
+
+    start_time = time.time()
+    diseases = load_entities(folder=bioconcepts_path, types=['disease'])[0]
```
Comment (emily-eai) [inline mat2vec/training/evaluate.py:113]: Are you assuming here that `disease` is at index 0?

File: mat2vec/training/evaluate.py [hunk 22]
```diff
@@ -0,0 +1,201 @@
...
+
+    w2v_model = Word2Vec.load(model_path)
+    entities1_found = 0
+    entities2_found = 0
+    unique_entities1 = set()
+    unique_entities2 = set()
+    pairs_found = 0
+    pairs_total = 0
+    recall = 0
+    MRR = 0
```
Comment (emily-eai) [inline mat2vec/training/evaluate.py:110]: It's not clear what this metric is. I assume it's often used in information retrieval. Is it Mean reciprocal rank?

File: mat2vec/training/evaluate.py [hunk 23]
```diff
@@ -0,0 +1,201 @@
...
+
+    start_time = time.time()
+    diseases = load_entities(folder=bioconcepts_path, types=['disease'])[0]
+    logging.info('Loaded entities {} in {}'.format(
+        len(diseases), datetime.timedelta(seconds=time.time() - start_time)))
+
+    with open(ref_pairs_path) as f:
+        for line in tqdm(f):
+            line = line.lower()
```
Comment (emily-eai) [inline mat2vec/training/evaluate.py:119]: Maybe show a sample line of what is being processed and also move all the processing to a helper function for clarity.

File: mat2vec/training/evaluate.py [hunk 24]
```diff
@@ -0,0 +1,201 @@
...
+            pairs_total += 1
+            if canonical_entity1:
+                entities1_found += 1
+            if canonical_entity2:
+                entities2_found += 1
+            if canonical_entity1 and canonical_entity2:
+                pairs_found += 1
+                word2vec_entity1 = canonical_entity1.replace(' ', '_')
+                rank = 0
+                for candidate_entity2, score in w2v_model.wv.most_similar(
```
Comment (emily-eai) [inline mat2vec/training/evaluate.py:144]: I don't think `score` is used anywhere. Also, `canonical` and `candidate` look very similar so it's quite confusing to read.

File: mat2vec/training/evaluate.py [hunk 25]
```diff
@@ -0,0 +1,201 @@
...
+                        word2vec_entity1, topn=1000 * k):
```
Comment (emily-eai) [inline mat2vec/training/evaluate.py:145]: Is topn different from topk?

File: mat2vec/training/evaluate.py [hunk 26]
```diff
@@ -0,0 +1,201 @@
...
+    pairs_total = 0
+    recall = 0
+    MRR = 0
+
+    start_time = time.time()
+    diseases = load_entities(folder=bioconcepts_path, types=['disease'])[0]
+    logging.info('Loaded entities {} in {}'.format(
+        len(diseases), datetime.timedelta(seconds=time.time() - start_time)))
+
+    with open(ref_pairs_path) as f:
```
Comment (emily-eai) [inline mat2vec/training/evaluate.py:117]: Perhaps add a documentation saying that you are only looking at gene-disease pairs for now, since the input arguments suggest this should work for any pair of entities but it is hard-coded here to parse gene-disease pairs.

File: mat2vec/training/evaluate.py [hunk 27]
```diff
@@ -0,0 +1,201 @@
...
+                entities2_found += 1
+            if canonical_entity1 and canonical_entity2:
+                pairs_found += 1
+                word2vec_entity1 = canonical_entity1.replace(' ', '_')
+                rank = 0
+                for candidate_entity2, score in w2v_model.wv.most_similar(
+                        word2vec_entity1, topn=1000 * k):
+                    candidate_entity2 = candidate_entity2.replace('_', ' ')
+                    if candidate_entity2 in diseases:
+                        rank += 1
```
Comment (emily-eai) [inline mat2vec/training/evaluate.py:148]: If I am understanding this correctly, the rank only increases if the candidate is found in the list of diseases? So if we have two models, one with two disease candidate entities at the top of the list, while the other has them at the bottom of the list, with other entity types before them, they would have the same rank?

File: mat2vec/training/helpers/utils.py (hunk 17 above)
Comment (emily-eai) [reply mat2vec/training/helpers/utils.py:34]: Would plural forms be automatically excluded too?

File: mat2vec/training/helpers/utils.py [hunk 28]
```diff
@@ -47,12 +67,12 @@ def keep_simple_formula(word, count, min_count):
 
 
 def compute_epoch_accuracies(root, prefix, analogy_file):
```
Comment (emily-eai) [inline mat2vec/training/helpers/utils.py:69]: I assume you haven't gotten around to documenting these yet.

File: mat2vec/training/phrase2vec.py [hunk 29]
```diff
@@ -2,30 +2,46 @@
...
+    """Given a list of words, excludes those from the keys of the phrase
+        dictionary.
+    """
     new_phrasergrams = {}
     words_re_list = []
     for word in words:
         we = regex.escape(word)
-        words_re_list.append("^" + we + "$|^" + we + "_|_" + we + "$|_" + we + "_")
-    word_reg = regex.compile(r""+"|".join(words_re_list))
+        words_re_list.append(
```
Comment (emily-eai) [inline mat2vec/training/phrase2vec.py:35]: Is this really more readable than the single line? Haha

File: mat2vec/training/phrase2vec.py [hunk 30]
```diff
@@ -2,30 +2,46 @@
...
+            "_")
+    word_reg = regex.compile(r"" + "|".join(words_re_list))
     for gram in tqdm(phrasegrams):
         valid = True
         for sub_gram in gram:
-            if word_reg.search(sub_gram.decode("unicode_escape", "ignore")) is not None:
+            if word_reg.search(sub_gram.decode(
+                    "unicode_escape", "ignore")) is not None:
                 valid = False
                 break
```
Comment (emily-eai) [inline mat2vec/training/phrase2vec.py:44]: Is this `break` for debugging? I am not exactly sure what is happening here, but it seems like `sub_gram` is a word-like token (multiple words?) and you are searching for the presence of these tokens within the words in the input list, and if found, makes the word invalid. In any case, it seems like the next lines would never be executed then since valid is ever only made False here?

File: mat2vec/training/phrase2vec.py [hunk 31]
```diff
@@ -57,64 +73,108 @@ def wordgrams(sent, depth, pc, th, ct, et, ip, d=0):
...
+        help="Folder for saving the model.")
     parser.add_argument("--epochs", default=30, help="Number of epochs.")
     parser.add_argument("--size", default=200, help="Size of the embedding.")
     parser.add_argument("--window", default=8, help="Context window size.")
-    parser.add_argument("--min_count", default=5, help="Minimum number of occurrences for word.")
+    parser.add_argument(
+        "--min_count",
+        default=5,
+        help="Minimum number of occurrences for word.")
     parser.add_argument("--workers", default=16, help="Number of workers.")
```
Comment (emily-eai) [inline mat2vec/training/phrase2vec.py:193]: Why are some split into multiple lines and some kept as a single line?

File: mat2vec/training/phrase2vec.py [hunk 32]
```diff
@@ -57,64 +73,108 @@ def wordgrams(sent, depth, pc, th, ct, et, ip, d=0):
...
+        default=0.0001,
+        help="Subsampling rate.")
+    parser.add_argument(
+        "--phrase_depth",
+        default=2,
+        help="The number of passes to perform for phrase generation.")
+    parser.add_argument(
+        "--phrase_count",
+        default=10,
+        help="Minimum number of occurrences for phrase to be considered.")
```
Comment (emily-eai) [inline mat2vec/training/phrase2vec.py:207]: Just wondering, why is the minimum number of phrase occurrences (10) higher than the minimum number of word occurrences (5)? Wouldn't it be more difficult for a phrase to occur, since it is a collection of words?

File: mat2vec/training/phrase2vec.py [hunk 33]
```diff
@@ -57,64 +73,108 @@ def wordgrams(sent, depth, pc, th, ct, et, ip, d=0):
...
+        "-notmp",
+        action="store_true",
+        help="If set, will not store the progress in tmp folder.")
     args = parser.parse_args()
 
     all_formula = []
     if args.keep_formula:
         try:
-            all_formula = load_obj(args.corpus + "_formula")  # list of formula is supplied
+            # list of formula is supplied
```
Comment (emily-eai) [inline mat2vec/training/phrase2vec.py]: Some of these code should probably be grouped and placed somewhere else.

**Review by emily-eai — COMMENTED** [review]
I have finished looking through all the code here and definitely skimmed some of the ones that are filled with jargon and had less documentation, so it was difficult to understand, but I tried to understand as much as I could and I hope the comments are helpful! I focused less on formatting etc. since it was generally fine and I assume some of the documentations are just not inserted as of yet, so I didn't point out everything.

File: mat2vec/processing/pubmed_to_lines.py (hunk 3 above)
Comment (rodrigo-eai) [reply mat2vec/processing/pubmed_to_lines.py:37]: Definitely, thanks!

File: mat2vec/processing/pubmed_to_lines.py (hunk 4 above)
Comment (rodrigo-eai) [reply mat2vec/processing/pubmed_to_lines.py:85]: Indeed, much better, thanks!

File: mat2vec/training/evaluate.py (hunk 5 above)
Comment (rodrigo-eai) [reply mat2vec/training/evaluate.py]: Sorry, the description is wrong. Thanks for noticing this!

File: mat2vec/training/evaluate.py (hunk 6 above)
Comment (rodrigo-eai) [reply mat2vec/training/evaluate.py:37]: This is was a hacky way to find the a match in the word2vec vocabulary. I think the right way to deal with these multiple entities is to use their average vector. I added a TODO note in the code.

File: mat2vec/training/evaluate.py (hunk 7 above)
Comment (rodrigo-eai) [reply mat2vec/training/evaluate.py]: Good point!

File: mat2vec/training/phrase2vec.py (hunk 29 above)
Comment (rodrigo-eai) [reply mat2vec/training/phrase2vec.py:35]: Hahah sorry, I blame autopep8 :D

File: mat2vec/training/phrase2vec.py (hunk 31 above)
Comment (rodrigo-eai) [reply mat2vec/training/phrase2vec.py:193]: Yeah, I blame autopep8 again (I though it would solve all my style formats but apparently it only handles well 90% of the cases)

File: mat2vec/training/phrase2vec.py (hunk 32 above)
Comment (rodrigo-eai) [reply mat2vec/training/phrase2vec.py:207]: That is a good point. I don't know either but I suggest keeping the default values as they are also used in the original word2vec implementation.

File: mat2vec/training/phrase2vec.py (hunk 30 above)
Comment (rodrigo-eai) [reply mat2vec/training/phrase2vec.py:44]: Thanks for paying attention to this. I'm also not sure what is happening here as I didn't wrote it. Since modifying this function might break something, I suggest to leave the logic the way it is. Also, we will probably use bio-related tokenizer in the future anyway.

File: mat2vec/training/phrase2vec.py (hunk 33 above)
Comment (rodrigo-eai) [reply mat2vec/training/phrase2vec.py]: Yeah, it is a mess. Thanks for suggesting this!

**Comment by rodrigo-eai** [comment]
@emily-eai, thanks for the thoughtful comments! Please take another look.

**Review by ghost — APPROVED** [review]
